cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -958,6 +987,71 @@ func getNotRegisteredNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances ma
 	return notRegistered
 }
 
+// Calculates which of the registered nodes in Kubernetes that do not exist in cloud provider.
+func (csr *ClusterStateRegistry) getCloudProviderDeletedNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances map[string][]cloudprovider.Instance) []*apiv1.Node {
+	nodesRemoved := make([]*apiv1.Node, 0)
+	currentCloudInstances := make(map[string]string, 0)
+	registeredNodes := make(map[string]*apiv1.Node, 0)
+	if len(allNodes) > 0 {
+		_, err := csr.cloudProvider.NodeExists(allNodes[0])
+		// Check if the cloud provider implements nodeExists method
+		nodeExistsNotImplemented := errors.Is(err, cloudprovider.ErrNotImplemented)
+		if nodeExistsNotImplemented {
+			// Fall-back to taint-based node deletion
+			for _, node := range allNodes {
+				if deletetaint.HasToBeDeletedTaint(node) {
+					nodesRemoved = append(nodesRemoved, node)
+				}
+			}
+		} else {
+			for nodeGroupName, instances := range cloudProviderNodeInstances {
+				for _, instance := range instances {
+					currentCloudInstances[instance.Id] = nodeGroupName
+				}
+			}
+			for _, node := range allNodes {
+				registeredNodes[node.Name] = node
+			}
+
+			// Fill previously deleted nodes, if they are still registered in Kubernetes
+			for nodeName, node := range csr.deletedNodes {
+				// Safety check to prevent flagging Kubernetes nodes as deleted
+				// if the Cloud Provider instance is re-discovered
+				_, cloudProviderFound := currentCloudInstances[node.Name]
+				if _, found := registeredNodes[nodeName]; found && !cloudProviderFound {
+					// Confirm that node is deleted by cloud provider, instead of
+					// a not-autoscaled node
+					nodeExists, existsErr := csr.cloudProvider.NodeExists(node)
+					if existsErr == nil && !nodeExists {
+						nodesRemoved = append(nodesRemoved, node)
+					}
+				}
+			}
+
+			// Seek nodes that may have been deleted since last update
+			// cloudProviderNodeInstances are retrieved by nodeGroup,
+			// not autoscaled nodes will be excluded
```

Comment:
How can we detect not autoscaled nodes that got deleted?

Reply:
I think the comment is confusing here, and apologize for that.  The `NodeGroupForNode` could be used to check whether the node is not-autoscaled.  Though I'm not sure that it is required to know if the node is not-autoscaled or not.  Should all nodes that no longer have a cloud instance be included in the deletion detection?

Reply:
Hm... actually considering only autoscaled nodes here is fine, I think counting upcoming nodes will work properly for autoscaled nodegroups only anyway.

Reply:
I agree, it should not impact the count since the readiness calculation considers only those nodes in autoscaled nodegroups.